src/mini_batch.jl

```diff
@@ -26,99 +33,100 @@ function kmeans!(alg::MiniBatch, X, k;
     N = zeros(T, k)
 
     # Initialize nearest centers for both batch and whole dataset labels
-    final_labels = Vector{Int}(undef, ncol)  # dataset labels
-
     converged = false
     niters = 0
     counter = 0
     J_previous = zero(T)
     J = zero(T)
+    totalcost = zero(T)
 
-    # TODO: Main Steps. Batch update centroids until convergence
+    # Main Steps. Batch update centroids until convergence
     while niters <= max_iters  # Step 4 in paper
 
         # b examples picked randomly from X (Step 5 in paper)
         batch_rand_idx = isnothing(weights) ? rand(rng, 1:ncol, alg.b) : wsample(rng, 1:ncol, weights, alg.b)
```

Comment:
`batch_rand_idx` should be put in a container (preallocated) and `rand!` version should be used instead of `rand`.

We should check `wsample`, I assume it has `wsample!` version as well

Reply:
both `rand!` and `wsample!` exist and drops allocations considerably. Nice catch as usual 👍🏾. `batch_rand_idx` pre-allocated in the next push.